### PR TITLE
Fixed #29087 -- Added delete buttons for unsaved admin inlines on validation error.

### PR DESCRIPTION
**Pull request by frnhr**
ticket: https://code.djangoproject.com/ticket/29087

Admin code, only templates and JS, no Python changes (except for selenium tests).

----

### Can delete failed inlines

(main intent of the PR)

Currently if an inline fails on form validation, it cannot be deleted / removed. The error has to be fixed, so that the inline can be saved. Only then can it be deleted. This can sometimes create situations where user has no choice but to reload the page, losing all data in the form:

Screenshot 2019-08-30 at 21 08 13
This PR adds the "X" delete buttons to all unsaved inlines, so in this case ☝️ the user needs only click the "X" next to offending inline and save.

### Can remove inlines to less then `extra = N`

(side-effect, IMHO positive although at first look it might seem negative)

#### Current state of things 

Setting `extra = 3` on the inline class (`3` is also the default) will show 3 empty inline forms without the delete buttons.

However, after saving those 3 inlines (provided there is any data to save), those 3 inlines get the `Delete` checkbox and can be deleted. So despite initial appearances, the `extra = 3` does not guarantee 3 related instances on the model. 

#### Changes in this PR

With this PR, those 3 inlines will have the delete buttons. Every inline will have either the delete button (for unsaved forms) or the delete checkbox (for saved forms). Permissions are respected as before.

Having the delete button on all unsaved inlines, (regardless of the `extra` value) is IMHO closer to the intended functionality of `extra` [as documented](https://docs.djangoproject.com/en/2.2/ref/contrib/admin/#django.contrib.admin.InlineModelAdmin.extra). Note that while there _is_ mention of inlines not exceeding `max_num` (a functionality which is kept in this PR), there is _no_ reference to not being able to go below `min_num` (the functionality which was quasi-provided with `extra = 3` and which is removed as a side-effect of this PR).

### CODE REVIEW

**Comment by frnhr** [comment]
Dear future reviewer, 

Even though this PR has `+400 -100`, the actual change to the code is rather small and simple. The rest is just moving some code around (packing it in a functions) and also tests.
I suggest taking a look at the PR in the "Ignore whitespace" mode as well:
Fixed__29087_--_Impossible_to_delete_pending_new_inlines_in_the_admin_after_unsuccessful_save_by_frnhr_·_Pull_Request__11504_·_django_django

**Comment by carltongibson** [comment]
(@frnhr I rebased on latest master, so at least we're up to date.)

**Comment by carltongibson** [comment]
I think we need to address for `TabularInline` as well here: 

Screenshot 2019-10-02 11 39 49

(With the CSS to match too. 🙂)

So, fix looks good. If you can make an initial "refactoring" commit, and then add the fix for the issue here, I think we won't be far off. 

I'm going to mark _Patch needs improvement_ on the ticket — please uncheck that when you're able to update, and optionally @carltongibson me here. 

Good work. 👍

**Comment by frnhr** [comment]
@carltongibson How does one do this @carltongibsoning you speak of? 😜 

I split the PR into 4 commits. Trivial to merge the last 3 commits, just say the word. Maybe this will help with the review a bit.

CSS changes are not necessary for this ticket. But I'd say they are related to it, and they seemed simple enough to include (just a bg color) so I did ¯\\\_(ツ)_/¯

**Comment by carltongibson** [comment]
@frnhr Super. Thanks for the hard work and quick turn around! I saw the commits coming in this afternoon. I shall review again in the morning. 🥇

**Comment by carltongibson** [comment]
buildbot, test on selenium.

**Comment by carltongibson** [comment]
buildbot, test on selenium.

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
`#11504 ` -> `#29087` in commits messages.

**Comment by carltongibson** [comment]
Ah, good. Well done @felixxm. That’s too much fighting with selenium that is! 🙂

**Comment by carltongibson** [comment]
OK, updated. (Sigh. Knocking off now. 🙂)

File: django/contrib/admin/static/admin/js/inlines.js [hunk 1]
```diff
@@ -289,7 +346,7 @@
                 $(selector).stackedFormset(selector, inlineOptions.options);
                 break;
             case "tabular":
-                selector = inlineOptions.name + "-group .tabular.inline-related tbody:first > tr";
```
[the remark this answers is not in the document]
Comment (frnhr) [reply django/contrib/admin/static/admin/js/inlines.js]: This prevents bogus behaviour when the first inline form has a non-field error(s). In tabular inline, those are rendered on their own TR. We don't want to add the "delete button" to that TR. 

In fewer words, prevents this "X":
Change_council___Django_site_admin

**Comment by carltongibson** [comment]
buildbot, test on selenium.

**Comment by carltongibson** [comment]
buildbot, test on selenium.

**Comment by felixxm** [comment]
buildbot, test on selenium.

**Comment by e-drake-p** [comment]
Apologies for what may be a silly question, but I've been poring through release notes and documentation and haven't yet been able to figure it out: What version of Django does this fix appear in? I have a user who would be happy to have this. Thank you!

**Comment by charettes** [comment]
@e-drake-p this is not part of any released version yet, it missed the 3.0 cutoff.

**Comment by e-drake-p** [comment]
Ah okay. Does that mean we will have to wait for 3.1 in August?

**Comment by frnhr** [comment]
@e-drake-p Seems so. I have created a branch on my fork which is based the latest 3.0.x plus my fix: https://github.com/frnhr/django/tree/3.0.x_w_29087 

Feel free to fork and use it, or if you feel particularly trusting, you can just install it directly from my fork: `pip install git+git://github.com/frnhr/django.git@3.0.x_w_29087#egg=django` (Might need to do `pip uninstall django` beforehand, not sure).

**Pull request merged**